dashboard/app/models/user.rb

```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
     if !script_sections.empty?
       # if we have one or more sections matching this script id, we consider a stage hidden if all of those sections
       # hides the stage
-      script_sections.all? {|s| script_level.stage_hidden_for_section?(s.id)}
+      script_sections.all? {|s| script_level.hidden_for_section?(s.id)}
     else
       # if we have no sections matching this script id, we consider a stage hidden if any of the sections we're in
       # hide it
-      sections.any? {|s| script_level.stage_hidden_for_section?(s.id)}
+      sections.any? {|s| script_level.hidden_for_section?(s.id)}
     end
   end
 
+  # This method will extra a list of hidden ids by section. The type of ids depends
+  # on the input. If hidden_stages is true, id is expected to be a script id and
+  # we look for stages that are hidden. If hidden_stages is false, id is expected
+  # to be a course_id, and we look for hidden scripts.
+  def get_hidden_ids(assign_id, hidden_stages = true)
+    # If we're a teacher, we want to go through each of our sections and return
+    # a mapping from section id to hidden stages in that section
+    if teacher?
+      hidden_by_section = {}
+      sections.each do |section|
+        if hidden_stages
+          # Given a script_id, looking for hidden stages
+          hidden_by_section[section.id] = section.section_hidden_stages.pluck(:stage_id)
+        else
+          # Given a course_id, looking for hidden_scripts
+          hidden_by_section[section.id] = section.section_hidden_scripts.pluck(:script_id)
+        end
```

Comment:
Can this be combined with the similar mapping below? Either in a helper called in both places, or collect the relevant sections for teacher / student, then map stage or script ids. They look very similar, but constructs a hash and the other uses a unique flat map. Does this mean the function returns either an array or hash depending on user type? Can they be consolidated? At least add a `@returns` function header explaining that.

Reply:
This function does return either an array or hash, which I've never really liked. Making this perhaps even sillier, the client - upon getting back an array - essentially turns it into a hash with a pretend section id of "STUDENT". https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/code-studio/hiddenStageRedux.js#L109

I could just as easily have the server always return a hash in this form. It would mean our fake STUDENT section id would now need to be shared between server/client, which feels gross, but maybe no grosser than what we have already.

Reply:
After starting to write this code, it still actually feels cleaner to me to have the server API return a different shape of data for teachers vs. students (and document this) then to have a magic section id that is not really a section id just to avoid this.

Reply:
I'm ok with that. Currently (without documentation), it's unexpected and confusing.